pkg/rtc/rtpstats.go

```diff
@@ -45,319 +90,273 @@ func (r *RTPStats) Stop() {
+
+		// update gap histogram
+		r.updateGapHistogram(int(diff))
+
+		// update missing sequence numbers
+		for lost := r.highestSN + 1; lost != rtp.SequenceNumber; lost++ {
+			r.clearSeenSN(lost)
+		}
+		r.packetsLost += uint32(diff - 1)
+
+		if rtp.SequenceNumber < r.highestSN {
+			r.cycles++
+		}
+		r.highestSN = rtp.SequenceNumber
+		r.highestTS = rtp.Timestamp
+		r.highestTime = packetTime
+
+		if rtp.Marker {
+			r.frames++
+		}
 	}
 
-	if aw := r.getActiveWindow(now); aw != nil {
-		aw.Update(rtp, packetTime)
+	// set current sequence number in seen list
+	r.setSeenSN(rtp.SequenceNumber)
+
+	if !isDuplicate {
+		r.updateJitter(rtp, packetTime)
 	}
+
+	return
 }
 
 func (r *RTPStats) UpdateNack(nackCount int, nackMissCount int) {
 	r.lock.Lock()
 	defer r.lock.Unlock()
 
-	now := time.Now()
-	if aw := r.getAggregateWindow(now); aw != nil {
-		aw.UpdateNack(nackCount, nackMissCount)
+	if !r.endTime.IsZero() {
+		return
 	}
 
-	if aw := r.getActiveWindow(now); aw != nil {
-		aw.UpdateNack(nackCount, nackMissCount)
-	}
+	r.nacks += uint32(nackCount)
+	r.nackMisses += uint32(nackMissCount)
 }
 
 func (r *RTPStats) UpdatePli() {
 	r.lock.Lock()
 	defer r.lock.Unlock()
 
-	now := time.Now()
-	if aw := r.getAggregateWindow(now); aw != nil {
-		aw.UpdatePli(now)
+	if !r.endTime.IsZero() {
+		return
 	}
 
-	if aw := r.getActiveWindow(now); aw != nil {
-		aw.UpdatePli(now)
-	}
+	r.plis++
+	r.lastPli = time.Now()
 }
 
-func (r *RTPStats) TimeSinceLastPliAggregate() int64 {
+func (r *RTPStats) TimeSinceLastPli() int64 {
 	r.lock.RLock()
 	defer r.lock.RUnlock()
 
-	now := time.Now()
-	if aw := r.getAggregateWindow(now); aw != nil {
-		return aw.TimeSinceLastPli(now)
-	}
-
-	return 0
+	return time.Now().UnixNano() - r.lastPli.UnixNano()
 }
 
-func (r *RTPStats) TimeSinceLastPliActive() int64 {
-	r.lock.RLock()
-	defer r.lock.RUnlock()
+func (r *RTPStats) UpdateFir() {
+	r.lock.Lock()
+	defer r.lock.Unlock()
 
-	now := time.Now()
-	if aw := r.getActiveWindow(now); aw != nil {
-		return aw.TimeSinceLastPli(now)
+	if !r.endTime.IsZero() {
+		return
 	}
 
-	return 0
+	r.firs++
+	r.lastFir = time.Now()
 }
 
-func (r *RTPStats) UpdateFir() {
+func (r *RTPStats) UpdateRtt(rtt uint32) {
 	r.lock.Lock()
 	defer r.lock.Unlock()
 
-	now := time.Now()
-	if aw := r.getAggregateWindow(now); aw != nil {
-		aw.UpdateFir(now)
+	if !r.endTime.IsZero() {
+		return
 	}
 
-	if aw := r.getActiveWindow(now); aw != nil {
-		aw.UpdateFir(now)
+	r.rtt = rtt
+	if r.rtt > r.maxRtt {
+		r.maxRtt = r.rtt
 	}
 }
 
-func (r *RTPStats) ToProtoActive() *livekit.RTPStats {
-	r.lock.RLock()
-	defer r.lock.RUnlock()
-
-	if r.activeWindow == nil {
-		return &livekit.RTPStats{}
-	}
+func (r *RTPStats) SetRtcpSenderReportData(rtpTS uint32, ntpTS buffer.NtpTime, arrival time.Time) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
 
-	return r.activeWindow.ToProto()
+	r.rtpSR = rtpTS
+	r.ntpSR = ntpTS
+	r.arrivalSR = arrival.UnixNano()
 }
 
-func (r *RTPStats) ToString() string {
+// RAJA-REMOVE-START
```

Comment:
I have this remove comment as I don't think we should be using this when creating the sender report for down tracks. Down track code is using the current layer and getting the time stamp. Different layers could have totally different time base. So, it cannot be used directly. There are two options
1. We should just use the down track timestamp (which is normalised across layers) OR
2. Lock to a layer of the published track and always use that layer's timestamp. This is a little more complicated.

For now, thinking of just using the down track timestamp. Note: have not started using the RTPStats yet in uptrack/downtrack. So, no changes yet there. Will be doing it in a subsequent PR.